src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -187,11 +188,67 @@ QVariant QgsSymbolV2LegendNode::data( int role ) const
         context.setRendererScale( scale );
         context.setMapToPixel( QgsMapToPixel( mupp ) ); // hope it's ok to leave out other params
 
-        pix = QgsSymbolLayerV2Utils::symbolPreviewPixmap( mItem.symbol(), iconSize, validData ? &context : 0 );
+        // crop
+        if ( mItem.symbol()->type() == QgsSymbolV2::Marker && mCrop )
+        {
+          pix = QgsSymbolLayerV2Utils::symbolPreviewPixmap( mItem.symbol(), QSize( 512, 512 ), validData ? &context : 0 );
```

Comment:
this section should be moved out to its own static function in QgsSymbolLayerV2Utils, it's a useful routine to have widely available